From af44d2728432ed45c63c66d72f6d8b0331e38dd2 Mon Sep 17 00:00:00 2001
From: Marius Gripsgard <marius@ubports.com>
Date: Sun, 6 Sep 2020 23:38:47 +0200
Subject: [PATCH] Bring back LanguagePluginInterface

Without this maliit-server is unable to cast the language plugin causing
a segfault rendering maliit-server non functional.

This partly reverts commit 1e5aa06d579fcd65fb6640b23b70318386ba3564.
---
 CMakeLists.txt                                |  5 +++-
 plugins/ar/src/arabicplugin.h                 |  3 +-
 plugins/az/src/azerbaijaniplugin.h            |  2 ++
 plugins/bs/src/bosnianplugin.h                |  2 ++
 plugins/ca/src/catalanplugin.h                |  2 ++
 plugins/chewing/src/chewingplugin.h           |  2 ++
 plugins/cs/src/czechplugin.h                  |  2 ++
 plugins/da/src/danishplugin.h                 |  2 ++
 plugins/de/src/germanplugin.h                 |  2 ++
 plugins/el/src/greekplugin.h                  |  2 ++
 plugins/emoji/src/emojiplugin.h               |  2 ++
 plugins/en/src/englishplugin.h                |  2 ++
 plugins/eo/src/esperantoplugin.h              |  2 ++
 plugins/es/src/spanishplugin.h                |  2 ++
 plugins/fa/src/persianplugin.h                |  2 ++
 plugins/fi/src/finnishplugin.h                |  2 ++
 plugins/fr/src/frenchplugin.h                 |  2 ++
 plugins/gd/src/gaelicplugin.h                 |  2 ++
 plugins/he/src/hebrewplugin.h                 |  2 ++
 plugins/hr/src/croatianplugin.h               |  2 ++
 plugins/hu/src/hungarianplugin.h              |  2 ++
 plugins/is/src/icelandicplugin.h              |  2 ++
 plugins/it/src/italianplugin.h                |  2 ++
 plugins/ja/src/japaneseplugin.h               |  2 ++
 plugins/ko/src/koreanplugin.h                 |  2 ++
 plugins/lv/src/latvianplugin.h                |  2 ++
 plugins/nb/src/norwegianplugin.h              |  2 ++
 plugins/nl/src/dutchplugin.h                  |  2 ++
 plugins/pinyin/src/pinyinplugin.h             |  2 ++
 plugins/pl/src/polishplugin.h                 |  2 ++
 plugins/pt/src/portugueseplugin.h             |  2 ++
 plugins/ro/src/romanianplugin.h               |  3 +-
 plugins/ru/src/russianplugin.h                |  2 ++
 plugins/sl/src/slovenianplugin.h              |  2 ++
 plugins/sr/src/serbianplugin.h                |  2 ++
 plugins/sv/src/swedishplugin.h                |  2 ++
 plugins/uk/src/ukrainianplugin.h              |  2 ++
 .../westernsupport/westernlanguagesplugin.h   |  2 ++
 src/lib/logic/abstractlanguageplugin.h        | 17 ++++++-----
 src/lib/logic/languageplugininterface.h       | 30 +++++++++++++++++++
 src/lib/logic/wordengine.cpp                  |  8 ++---
 src/lib/logic/wordengine.h                    |  1 +
 tests/testlayout/src/testlayoutplugin.h       |  2 ++
 43 files changed, 124 insertions(+), 15 deletions(-)
 create mode 100644 src/lib/logic/languageplugininterface.h

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7f3f366c..ec39c1a5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -48,10 +48,12 @@ set(MALIIT_KEYBOARD_LIB_SOURCES
         src/lib/logic/abstractwordengine.h
         src/lib/logic/eventhandler.cpp
         src/lib/logic/eventhandler.h
+        src/lib/logic/languageplugininterface.h
         src/lib/logic/style.cpp
         src/lib/logic/style.h
         src/lib/logic/wordengine.cpp
         src/lib/logic/wordengine.h
+
         src/lib/models/area.cpp
         src/lib/models/area.h
         src/lib/models/key.cpp
@@ -70,6 +72,7 @@ set(MALIIT_KEYBOARD_LIB_SOURCES
         src/lib/models/wordcandidate.h
         src/lib/models/wordribbon.cpp
         src/lib/models/wordribbon.h
+
         src/lib/coreutils.cpp
         src/lib/coreutils.h)
 
@@ -86,7 +89,7 @@ set(WESTERNSUPPORT_SOURCES
         src/lib/logic/abstractlanguageplugin.cpp
         src/lib/logic/abstractlanguageplugin.h)
 
-# TODO install logic/abstractplugininterface.h as HEADERS
+# TODO install logic/languageplugininterface.h and logic/abstractplugininterface.h as HEADERS
 
 set(maliit-keyboard-libraries Qt5::Core)
 set(maliit-keyboard-definitions HUNSPELL_DICT_PATH="${HUNSPELL_DICT_PATH}"
diff --git a/plugins/ar/src/arabicplugin.h b/plugins/ar/src/arabicplugin.h
index ec05949a..034ca226 100644
--- a/plugins/ar/src/arabicplugin.h
+++ b/plugins/ar/src/arabicplugin.h
@@ -2,13 +2,14 @@
 #define ARABICPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class ArabicPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "arabicplugin.json")
-
+    Q_INTERFACES(LanguagePluginInterface)
 public:
     explicit ArabicPlugin(QObject* parent = nullptr)
         : WesternLanguagesPlugin(parent)
diff --git a/plugins/az/src/azerbaijaniplugin.h b/plugins/az/src/azerbaijaniplugin.h
index e44c25d5..7a198fd3 100644
--- a/plugins/az/src/azerbaijaniplugin.h
+++ b/plugins/az/src/azerbaijaniplugin.h
@@ -2,6 +2,7 @@
 #define AZERBAIJANIPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 //#include <presage.h>
@@ -10,6 +11,7 @@ class AzerbaijaniPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "azerbaijaniplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit AzerbaijaniPlugin(QObject* parent = nullptr)
diff --git a/plugins/bs/src/bosnianplugin.h b/plugins/bs/src/bosnianplugin.h
index a3667b38..db8c2d17 100644
--- a/plugins/bs/src/bosnianplugin.h
+++ b/plugins/bs/src/bosnianplugin.h
@@ -2,12 +2,14 @@
 #define BOSNIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class BosnianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "bosnianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit BosnianPlugin(QObject* parent = nullptr)
diff --git a/plugins/ca/src/catalanplugin.h b/plugins/ca/src/catalanplugin.h
index a57bfadc..20672efe 100644
--- a/plugins/ca/src/catalanplugin.h
+++ b/plugins/ca/src/catalanplugin.h
@@ -2,12 +2,14 @@
 #define CATALANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class CatalanPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "catalanplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit CatalanPlugin(QObject* parent = nullptr)
diff --git a/plugins/chewing/src/chewingplugin.h b/plugins/chewing/src/chewingplugin.h
index dd4c939b..5c5bbb74 100644
--- a/plugins/chewing/src/chewingplugin.h
+++ b/plugins/chewing/src/chewingplugin.h
@@ -4,6 +4,7 @@
 #include <QObject>
 #include <QStringList>
 #include <QThread>
+#include "languageplugininterface.h"
 #include "abstractlanguageplugin.h"
 
 #include "chewingadapter.h"
@@ -15,6 +16,7 @@ class ChewingPlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "chewingplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit ChewingPlugin(QObject *parent = nullptr);
diff --git a/plugins/cs/src/czechplugin.h b/plugins/cs/src/czechplugin.h
index c6228fb8..a797c84d 100644
--- a/plugins/cs/src/czechplugin.h
+++ b/plugins/cs/src/czechplugin.h
@@ -2,12 +2,14 @@
 #define CZECHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class CzechPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "czechplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit CzechPlugin(QObject* parent = nullptr)
diff --git a/plugins/da/src/danishplugin.h b/plugins/da/src/danishplugin.h
index 2ed0a8df..83383c39 100644
--- a/plugins/da/src/danishplugin.h
+++ b/plugins/da/src/danishplugin.h
@@ -2,12 +2,14 @@
 #define DANISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class DanishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "danishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit DanishPlugin(QObject* parent = nullptr)
diff --git a/plugins/de/src/germanplugin.h b/plugins/de/src/germanplugin.h
index 3ac7e5db..0453d6b0 100644
--- a/plugins/de/src/germanplugin.h
+++ b/plugins/de/src/germanplugin.h
@@ -2,12 +2,14 @@
 #define GERMANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class GermanPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "germanplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit GermanPlugin(QObject* parent = nullptr)
diff --git a/plugins/el/src/greekplugin.h b/plugins/el/src/greekplugin.h
index cdc69e43..b1ac71ff 100644
--- a/plugins/el/src/greekplugin.h
+++ b/plugins/el/src/greekplugin.h
@@ -2,12 +2,14 @@
 #define GREEKPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class GreekPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "greekplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit GreekPlugin(QObject* parent = nullptr)
diff --git a/plugins/emoji/src/emojiplugin.h b/plugins/emoji/src/emojiplugin.h
index 304ea910..37b2b9e6 100644
--- a/plugins/emoji/src/emojiplugin.h
+++ b/plugins/emoji/src/emojiplugin.h
@@ -3,6 +3,7 @@
 
 #include <QObject>
 #include <QStringList>
+#include "languageplugininterface.h"
 #include "abstractlanguageplugin.h"
 
 #include <iostream>
@@ -13,6 +14,7 @@ class EmojiPlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "emojiplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit EmojiPlugin(QObject *parent = nullptr);
diff --git a/plugins/en/src/englishplugin.h b/plugins/en/src/englishplugin.h
index e1761a26..b8c35dd7 100644
--- a/plugins/en/src/englishplugin.h
+++ b/plugins/en/src/englishplugin.h
@@ -2,6 +2,7 @@
 #define ENGLISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 //#include <presage.h>
@@ -10,6 +11,7 @@ class EnglishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "englishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit EnglishPlugin(QObject* parent = nullptr)
diff --git a/plugins/eo/src/esperantoplugin.h b/plugins/eo/src/esperantoplugin.h
index aaff5b2e..79607f90 100644
--- a/plugins/eo/src/esperantoplugin.h
+++ b/plugins/eo/src/esperantoplugin.h
@@ -2,12 +2,14 @@
 #define ESPERANTOPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class EsperantoPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "esperantoplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit EsperantoPlugin(QObject* parent = nullptr)
diff --git a/plugins/es/src/spanishplugin.h b/plugins/es/src/spanishplugin.h
index cb332a7f..6ffdc3cf 100644
--- a/plugins/es/src/spanishplugin.h
+++ b/plugins/es/src/spanishplugin.h
@@ -2,12 +2,14 @@
 #define SPANISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class SpanishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "spanishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit SpanishPlugin(QObject* parent = nullptr)
diff --git a/plugins/fa/src/persianplugin.h b/plugins/fa/src/persianplugin.h
index a8e1afb8..60807509 100644
--- a/plugins/fa/src/persianplugin.h
+++ b/plugins/fa/src/persianplugin.h
@@ -2,12 +2,14 @@
 #define PERSIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class PersianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "persianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit PersianPlugin(QObject* parent = nullptr)
diff --git a/plugins/fi/src/finnishplugin.h b/plugins/fi/src/finnishplugin.h
index 9cc9ee83..9d564b48 100644
--- a/plugins/fi/src/finnishplugin.h
+++ b/plugins/fi/src/finnishplugin.h
@@ -2,12 +2,14 @@
 #define FINNISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class FinnishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "finnishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit FinnishPlugin(QObject* parent = nullptr)
diff --git a/plugins/fr/src/frenchplugin.h b/plugins/fr/src/frenchplugin.h
index 2ec4e2db..66d9fe7f 100644
--- a/plugins/fr/src/frenchplugin.h
+++ b/plugins/fr/src/frenchplugin.h
@@ -2,6 +2,7 @@
 #define FRENCHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 //#include <presage.h>
@@ -10,6 +11,7 @@ class FrenchPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "frenchplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit FrenchPlugin(QObject* parent = nullptr)
diff --git a/plugins/gd/src/gaelicplugin.h b/plugins/gd/src/gaelicplugin.h
index 0f9d40d5..ba724453 100644
--- a/plugins/gd/src/gaelicplugin.h
+++ b/plugins/gd/src/gaelicplugin.h
@@ -2,6 +2,7 @@
 #define GAELICPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 //#include <presage.h>
@@ -10,6 +11,7 @@ class GaelicPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "gaelicplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit GaelicPlugin(QObject* parent = nullptr)
diff --git a/plugins/he/src/hebrewplugin.h b/plugins/he/src/hebrewplugin.h
index 55ec386c..92ce1107 100644
--- a/plugins/he/src/hebrewplugin.h
+++ b/plugins/he/src/hebrewplugin.h
@@ -2,12 +2,14 @@
 #define HEBREWPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class HebrewPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "hebrewplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit HebrewPlugin(QObject* parent = nullptr)
diff --git a/plugins/hr/src/croatianplugin.h b/plugins/hr/src/croatianplugin.h
index bc207997..3a8d508d 100644
--- a/plugins/hr/src/croatianplugin.h
+++ b/plugins/hr/src/croatianplugin.h
@@ -2,12 +2,14 @@
 #define CROATIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class CroatianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "croatianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit CroatianPlugin(QObject* parent = nullptr)
diff --git a/plugins/hu/src/hungarianplugin.h b/plugins/hu/src/hungarianplugin.h
index 6984281e..1bc882b7 100644
--- a/plugins/hu/src/hungarianplugin.h
+++ b/plugins/hu/src/hungarianplugin.h
@@ -2,12 +2,14 @@
 #define HUNGARIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class HungarianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "hungarianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit HungarianPlugin(QObject* parent = nullptr)
diff --git a/plugins/is/src/icelandicplugin.h b/plugins/is/src/icelandicplugin.h
index 7ae0bb1b..40f670d6 100644
--- a/plugins/is/src/icelandicplugin.h
+++ b/plugins/is/src/icelandicplugin.h
@@ -2,12 +2,14 @@
 #define ICELANDICPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class IcelandicPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "icelandicplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit IcelandicPlugin(QObject* parent = nullptr)
diff --git a/plugins/it/src/italianplugin.h b/plugins/it/src/italianplugin.h
index a9d6dfe2..91fe3c85 100644
--- a/plugins/it/src/italianplugin.h
+++ b/plugins/it/src/italianplugin.h
@@ -2,12 +2,14 @@
 #define ITALIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class ItalianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "italianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit ItalianPlugin(QObject* parent = nullptr)
diff --git a/plugins/ja/src/japaneseplugin.h b/plugins/ja/src/japaneseplugin.h
index e934dbb6..fc583821 100644
--- a/plugins/ja/src/japaneseplugin.h
+++ b/plugins/ja/src/japaneseplugin.h
@@ -3,6 +3,7 @@
 
 #include <QObject>
 #include <QThread>
+#include "languageplugininterface.h"
 #include "abstractlanguageplugin.h"
 
 #include "anthyadapter.h"
@@ -13,6 +14,7 @@ class JapanesePlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "japaneseplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit JapanesePlugin(QObject* parent = nullptr);
diff --git a/plugins/ko/src/koreanplugin.h b/plugins/ko/src/koreanplugin.h
index 10bf7cc8..20e3ae6c 100644
--- a/plugins/ko/src/koreanplugin.h
+++ b/plugins/ko/src/koreanplugin.h
@@ -3,6 +3,7 @@
 
 #include <QObject>
 #include <QStringList>
+#include "languageplugininterface.h"
 #include "abstractlanguageplugin.h"
 #include "candidatescallback.h"
 #include "spellchecker.h"
@@ -17,6 +18,7 @@ class KoreanPlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "koreanplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit KoreanPlugin(QObject *parent = nullptr);
diff --git a/plugins/lv/src/latvianplugin.h b/plugins/lv/src/latvianplugin.h
index 44c6fbc0..836272ca 100644
--- a/plugins/lv/src/latvianplugin.h
+++ b/plugins/lv/src/latvianplugin.h
@@ -2,12 +2,14 @@
 #define LATVIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class LatvianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "latvianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit LatvianPlugin(QObject* parent = nullptr)
diff --git a/plugins/nb/src/norwegianplugin.h b/plugins/nb/src/norwegianplugin.h
index 4efa5e25..2e28314c 100644
--- a/plugins/nb/src/norwegianplugin.h
+++ b/plugins/nb/src/norwegianplugin.h
@@ -2,12 +2,14 @@
 #define NORWEGIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class NorwegianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "norwegianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit NorwegianPlugin(QObject* parent = nullptr)
diff --git a/plugins/nl/src/dutchplugin.h b/plugins/nl/src/dutchplugin.h
index 5b42ef00..c4be7697 100644
--- a/plugins/nl/src/dutchplugin.h
+++ b/plugins/nl/src/dutchplugin.h
@@ -2,12 +2,14 @@
 #define DUTCHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class DutchPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "dutchplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit DutchPlugin(QObject* parent = nullptr)
diff --git a/plugins/pinyin/src/pinyinplugin.h b/plugins/pinyin/src/pinyinplugin.h
index 52306458..dc640ea5 100644
--- a/plugins/pinyin/src/pinyinplugin.h
+++ b/plugins/pinyin/src/pinyinplugin.h
@@ -4,6 +4,7 @@
 #include <QObject>
 #include <QStringList>
 #include <QThread>
+#include "languageplugininterface.h"
 #include "abstractlanguageplugin.h"
 
 #include "pinyinadapter.h"
@@ -15,6 +16,7 @@ class PinyinPlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "pinyinplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit PinyinPlugin(QObject *parent = nullptr);
diff --git a/plugins/pl/src/polishplugin.h b/plugins/pl/src/polishplugin.h
index 4c76b009..4f29d29c 100644
--- a/plugins/pl/src/polishplugin.h
+++ b/plugins/pl/src/polishplugin.h
@@ -2,12 +2,14 @@
 #define POLISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class PolishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "polishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit PolishPlugin(QObject* parent = nullptr)
diff --git a/plugins/pt/src/portugueseplugin.h b/plugins/pt/src/portugueseplugin.h
index 9401d146..6f87f5b5 100644
--- a/plugins/pt/src/portugueseplugin.h
+++ b/plugins/pt/src/portugueseplugin.h
@@ -2,12 +2,14 @@
 #define PORTUGUESEPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class PortuguesePlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "portugueseplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit PortuguesePlugin(QObject* parent = nullptr)
diff --git a/plugins/ro/src/romanianplugin.h b/plugins/ro/src/romanianplugin.h
index c05ffc27..4f76348d 100644
--- a/plugins/ro/src/romanianplugin.h
+++ b/plugins/ro/src/romanianplugin.h
@@ -2,12 +2,14 @@
 #define ROMANIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class RomanianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "romanianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit RomanianPlugin(QObject* parent = nullptr)
@@ -19,4 +21,3 @@ class RomanianPlugin : public WesternLanguagesPlugin
 };
 
 #endif // ROMANIANPLUGIN_H
-
diff --git a/plugins/ru/src/russianplugin.h b/plugins/ru/src/russianplugin.h
index bbf6a8f1..2d51f610 100644
--- a/plugins/ru/src/russianplugin.h
+++ b/plugins/ru/src/russianplugin.h
@@ -2,12 +2,14 @@
 #define RUSSIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class RussianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "russianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit RussianPlugin(QObject* parent = nullptr)
diff --git a/plugins/sl/src/slovenianplugin.h b/plugins/sl/src/slovenianplugin.h
index 763b0eb9..86e8c2d5 100644
--- a/plugins/sl/src/slovenianplugin.h
+++ b/plugins/sl/src/slovenianplugin.h
@@ -2,12 +2,14 @@
 #define SLOVENIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class SlovenianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "slovenianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit SlovenianPlugin(QObject* parent = nullptr)
diff --git a/plugins/sr/src/serbianplugin.h b/plugins/sr/src/serbianplugin.h
index aaecf774..146639e6 100644
--- a/plugins/sr/src/serbianplugin.h
+++ b/plugins/sr/src/serbianplugin.h
@@ -2,12 +2,14 @@
 #define SERBIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class SerbianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "serbianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit SerbianPlugin(QObject* parent = nullptr)
diff --git a/plugins/sv/src/swedishplugin.h b/plugins/sv/src/swedishplugin.h
index 73f09d7c..5d2b374f 100644
--- a/plugins/sv/src/swedishplugin.h
+++ b/plugins/sv/src/swedishplugin.h
@@ -2,12 +2,14 @@
 #define SWEDISHPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class SwedishPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "swedishplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit SwedishPlugin(QObject* parent = nullptr)
diff --git a/plugins/uk/src/ukrainianplugin.h b/plugins/uk/src/ukrainianplugin.h
index 0d8abade..9c063761 100644
--- a/plugins/uk/src/ukrainianplugin.h
+++ b/plugins/uk/src/ukrainianplugin.h
@@ -2,12 +2,14 @@
 #define UKRAINIANPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class UkrainianPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "ukrainianplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit UkrainianPlugin(QObject* parent = nullptr)
diff --git a/plugins/westernsupport/westernlanguagesplugin.h b/plugins/westernsupport/westernlanguagesplugin.h
index ee103e87..88fe9939 100644
--- a/plugins/westernsupport/westernlanguagesplugin.h
+++ b/plugins/westernsupport/westernlanguagesplugin.h
@@ -3,6 +3,7 @@
 
 #include <QObject>
 
+#include "languageplugininterface.h"
 #include "westernlanguagefeatures.h"
 #include "spellchecker.h"
 #include "abstractlanguageplugin.h"
@@ -18,6 +19,7 @@ class SpellPredictWorker;
 class WesternLanguagesPlugin : public AbstractLanguagePlugin
 {
     Q_OBJECT
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit WesternLanguagesPlugin(QObject *parent = nullptr);
diff --git a/src/lib/logic/abstractlanguageplugin.h b/src/lib/logic/abstractlanguageplugin.h
index a1d393d9..0b1a886d 100644
--- a/src/lib/logic/abstractlanguageplugin.h
+++ b/src/lib/logic/abstractlanguageplugin.h
@@ -30,24 +30,25 @@
 
 #include <QObject>
 
-class AbstractLanguageFeatures;
+#include "languageplugininterface.h"
 
-class AbstractLanguagePlugin : public QObject
+class AbstractLanguagePlugin : public QObject, public LanguagePluginInterface
 {
     Q_OBJECT
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     AbstractLanguagePlugin(QObject *parent = nullptr);
     ~AbstractLanguagePlugin() override;
 
-    virtual void predict(const QString& surroundingLeft, const QString& preedit);
-    virtual void wordCandidateSelected(QString word);
-    virtual AbstractLanguageFeatures* languageFeature();
+    void predict(const QString& surroundingLeft, const QString& preedit) override;
+    void wordCandidateSelected(QString word) override;
+    AbstractLanguageFeatures* languageFeature() override;
 
     //! spell checker
-    virtual void spellCheckerSuggest(const QString& word, int limit);
-    virtual void addToSpellCheckerUserWordList(const QString& word);
-    virtual bool setLanguage(const QString& languageId, const QString& pluginPath);
+    void spellCheckerSuggest(const QString& word, int limit) override;
+    void addToSpellCheckerUserWordList(const QString& word) override;
+    bool setLanguage(const QString& languageId, const QString& pluginPath) override;
 
 signals:
     void newSpellingSuggestions(QString word, QStringList suggestions);
diff --git a/src/lib/logic/languageplugininterface.h b/src/lib/logic/languageplugininterface.h
new file mode 100644
index 00000000..2de89ba1
--- /dev/null
+++ b/src/lib/logic/languageplugininterface.h
@@ -0,0 +1,30 @@
+#ifndef LANGUAGEPLUGININTERFACE_H
+#define LANGUAGEPLUGININTERFACE_H
+
+#include <QString>
+#include <QStringList>
+
+class AbstractLanguageFeatures;
+
+class LanguagePluginInterface
+{
+public:
+    virtual ~LanguagePluginInterface() = default;
+
+    virtual void predict(const QString& surroundingLeft, const QString& preedit) = 0;
+    virtual void wordCandidateSelected(QString word) = 0;
+
+    virtual AbstractLanguageFeatures* languageFeature() = 0;
+
+    //! spell checker
+    virtual void spellCheckerSuggest(const QString& word, int limit) = 0;
+    virtual void addToSpellCheckerUserWordList(const QString& word) = 0;
+    virtual bool setLanguage(const QString& languageId, const QString &pluginPath) = 0;
+};
+
+#define LanguagePluginInterface_iid "com.canonical.UbuntuKeyboard.LanguagePluginInterface"
+
+Q_DECLARE_INTERFACE(LanguagePluginInterface, LanguagePluginInterface_iid)
+
+#endif // LANGUAGEPLUGININTERFACE_H
+
diff --git a/src/lib/logic/wordengine.cpp b/src/lib/logic/wordengine.cpp
index 47bc1b4c..16bf91e5 100644
--- a/src/lib/logic/wordengine.cpp
+++ b/src/lib/logic/wordengine.cpp
@@ -60,7 +60,7 @@ class WordEnginePrivate
 
     bool clear_candidates_on_incoming;
 
-    AbstractLanguagePlugin* languagePlugin;
+    LanguagePluginInterface* languagePlugin;
 
     QPluginLoader pluginLoader;
 
@@ -94,7 +94,7 @@ class WordEnginePrivate
         QObject *plugin = pluginLoader.instance();
 
         if (plugin) {
-            languagePlugin = qobject_cast<AbstractLanguagePlugin *>(plugin);
+            languagePlugin = qobject_cast<LanguagePluginInterface *>(plugin);
             if (!languagePlugin) {
                 qCritical() << "wordengine.cpp - loading plugin failed: " + pluginPath;
 
@@ -411,9 +411,9 @@ void WordEngine::onLanguageChanged(const QString &pluginPath, const QString &lan
 
     Q_EMIT enabledChanged(isEnabled());
 
-    connect(d->languagePlugin, &AbstractLanguagePlugin::newSpellingSuggestions,
+    connect(static_cast<AbstractLanguagePlugin *>(d->languagePlugin), &AbstractLanguagePlugin::newSpellingSuggestions,
             this, &WordEngine::newSpellingSuggestions);
-    connect(d->languagePlugin, &AbstractLanguagePlugin::newPredictionSuggestions,
+    connect(static_cast<AbstractLanguagePlugin *>(d->languagePlugin), &AbstractLanguagePlugin::newPredictionSuggestions,
             this, &WordEngine::newPredictionSuggestions);
     Q_EMIT pluginChanged();
 }
diff --git a/src/lib/logic/wordengine.h b/src/lib/logic/wordengine.h
index 2c28a11f..233a1ff1 100644
--- a/src/lib/logic/wordengine.h
+++ b/src/lib/logic/wordengine.h
@@ -35,6 +35,7 @@
 #include "models/text.h"
 #include "logic/abstractwordengine.h"
 #include "logic/abstractlanguagefeatures.h"
+#include "languageplugininterface.h"
 
 #include <QtCore>
 #include <QMutex>
diff --git a/tests/testlayout/src/testlayoutplugin.h b/tests/testlayout/src/testlayoutplugin.h
index 602c5043..df0c21c4 100644
--- a/tests/testlayout/src/testlayoutplugin.h
+++ b/tests/testlayout/src/testlayoutplugin.h
@@ -2,12 +2,14 @@
 #define TESTLAYOUTPLUGIN_H
 
 #include <QObject>
+#include "languageplugininterface.h"
 #include "westernlanguagesplugin.h"
 
 class TestLayoutPlugin : public WesternLanguagesPlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "io.maliit.keyboard.LanguagePlugin.1" FILE "testlayoutplugin.json")
+    Q_INTERFACES(LanguagePluginInterface)
 
 public:
     explicit TestLayoutPlugin(QObject* parent = 0)

